File: backend-toolrent/src/main/java/com/example/demo/Service/KardexService.java
package com.example.demo.Service;

import com.example.demo.Entity.KardexEntity;
import com.example.demo.Entity.ToolEntity;
import com.example.demo.Repository.KardexRepository;
import com.example.demo.Repository.ToolRepository;
import jakarta.transaction.Transactional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.List;

@Service
public class KardexService {

    @Autowired
    private KardexRepository kardexRepository;

    @Autowired
    private ToolRepository toolRepository;

    /**
     * RF5.1: Registrar movimiento en el kardex
     */
    @Transactional
    public KardexEntity registerMovement(
            Long toolId,
            String movementType,
            Integer quantity,
            String username,
            String observations,
            Long loanId
    ) {
        // Validaciones
        if (toolId == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "El ID de herramienta es obligatorio");
        }
        if (movementType == null || movementType.isBlank()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "El tipo de movimiento es obligatorio");
        }
        if (quantity == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "La cantidad es obligatoria");
        }

        // Buscar herramienta
        ToolEntity tool = toolRepository.findById(toolId)
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND,
                        "Herramienta no encontrada con ID: " + toolId
                ));

        // Crear movimiento
        KardexEntity kardex = new KardexEntity();
        kardex.setMovementType(movementType.toUpperCase());
        kardex.setTool(tool);
        kardex.setQuantity(quantity);
        kardex.setUsername(username != null ? username : "SYSTEM");
        kardex.setMovementDate(LocalDateTime.now());
        kardex.setObservations(observations);
        kardex.setLoanId(loanId);

        return kardexRepository.save(kardex);
    }

    /**
     * RF5.2: Consultar historial de movimientos por herramienta
     */
    public List<KardexEntity> getMovementsByTool(Long toolId) {
        if (toolId == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "El ID de herramienta es obligatorio");
        }

        ToolEntity tool = toolRepository.findById(toolId)
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND,
                        "Herramienta no encontrada con ID: " + toolId
                ));

        return kardexRepository.findByToolOrderByMovementDateDesc(tool);
    }

    /**
     * RF5.3: Generar listado de movimientos por rango de fechas
     */
    public List<KardexEntity> getMovementsByDateRange(LocalDate startDate, LocalDate endDate) {
        if (startDate == null || endDate == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Las fechas son obligatorias");
        }

        if (startDate.isAfter(endDate)) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "La fecha de inicio debe ser anterior a la fecha de fin");
        }

        LocalDateTime startDateTime = startDate.atStartOfDay();
        LocalDateTime endDateTime = endDate.atTime(LocalTime.MAX);

        return kardexRepository.findByMovementDateBetweenOrderByMovementDateDesc(startDateTime, endDateTime);
    }

    /**
     * Obtener todos los movimientos
     */
    public List<KardexEntity> getAllMovements() {
        return kardexRepository.findAllByOrderByMovementDateDesc();
    }

    /**
     * Obtener movimientos por tipo
     */
    public List<KardexEntity> getMovementsByType(String movementType) {
        if (movementType == null || movementType.isBlank()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "El tipo de movimiento es obligatorio");
        }
        return kardexRepository.findByMovementTypeOrderByMovementDateDesc(movementType.toUpperCase());
    }
}

File: toolrent-frontend/src/services/kardex.service.js
// src/services/kardex.service.js
import http from '../http-common'

/**
 * Servicio para gestionar el Kardex (historial de movimientos)
 * Épica 5: Gestión de Kardex y Movimientos
 */

/**
 * RF5.2: Consultar el historial de movimientos de una herramienta específica
 * GET /api/v1/kardex/tool/{toolId}
 */
const getMovementsByTool = async (toolId) => {
  try {
    const { data } = await http.get(`/api/v1/kardex/tool/${toolId}`)
    return Array.isArray(data) ? data : []
  } catch (error) {
    console.error('Error al obtener movimientos por herramienta:', error)
    throw error
  }
}

/**
 * RF5.3: Consultar movimientos en un rango de fechas
 * GET /api/v1/kardex/date-range?startDate={startDate}&endDate={endDate}
 */
const getMovementsByDateRange = async (startDate, endDate) => {
  try {
    const { data } = await http.get('/api/v1/kardex/date-range', {
      params: { 
        startDate: startDate,
        endDate: endDate 
      }
    })
    return Array.isArray(data) ? data : []
  } catch (error) {
    console.error('Error al obtener movimientos por rango de fechas:', error)
    throw error
  }
}

/**
 * Obtener todos los movimientos del kardex
 * GET /api/v1/kardex
 */
const getAllMovements = async () => {
  try {
    const { data } = await http.get('/api/v1/kardex')
    return Array.isArray(data) ? data : []
  } catch (error) {
    console.error('Error al obtener todos los movimientos:', error)
    throw error
  }
}

/**
 * Filtrar movimientos por tipo
 * GET /api/v1/kardex/type/{movementType}
 * Tipos válidos: INGRESO, PRESTAMO, DEVOLUCION, BAJA, REPARACION
 */
const getMovementsByType = async (movementType) => {
  try {
    const { data } = await http.get(`/api/v1/kardex/type/${movementType}`)
    return Array.isArray(data) ? data : []
  } catch (error) {
    console.error('Error al obtener movimientos por tipo:', error)
    throw error
  }
}

const kardexService = {
  getMovementsByTool,
  getMovementsByDateRange,
  getAllMovements,
  getMovementsByType,
}

export default kardexService
